### PR TITLE
Reworked driver exceptions

### PR DESCRIPTION
**Pull request by morozov**
|      Q       |   A
|------------- | -----------
| Type         | improvement
| BC Break     | yes

Besides the breaking changes necessitated by enforcing strict scalar types and adding return type hints, named constructors were added to `DB2Exception` and `MysqliException`.

### CODE REVIEW

File: lib/Doctrine/DBAL/Driver/DriverException.php [hunk 1]
```diff
@@ -14,29 +14,10 @@
...
-
     /**
      * Returns the SQLSTATE the driver was in at the time the error occurred.
      *
      * Returns null if the driver does not provide a SQLSTATE for the error occurred.
-     *
-     * @return string|null
      */
-    public function getSQLState();
+    public function getSQLState() : ?string;
```
Comment (greg0ire) [inline lib/Doctrine/DBAL/Driver/DriverException.php]: Shouldn't this BC-break be listed too?

Comment (morozov) [reply lib/Doctrine/DBAL/Driver/DriverException.php]: Why is it a BC break?

Comment (Ocramius) [reply lib/Doctrine/DBAL/Driver/DriverException.php]: Indeed, not a BC break: https://3v4l.org/uvKF4

**Review by Ocramius — APPROVED** [review]
:+1:

File: lib/Doctrine/DBAL/Driver/DriverException.php (hunk 1 above)
Comment (Majkl578) [reply lib/Doctrine/DBAL/Driver/DriverException.php]: It's a BC break for those who implemented that without typehint, i.e. in mocks.

File: UPGRADE.md [hunk 2]
```diff
@@ -1,5 +1,13 @@
 # Upgrade to 3.0
 
+## BC BREAK Changes in driver exceptions
+
+1. The `Doctrine\DBAL\Driver\DriverException::getErrorCode()` method is removed. In order to obtain the driver error code, please use `::getCode()`.
+2. `Doctrine\DBAL\Driver\PDOException` no longer extends `PDOException`.
+3. The value returned by `Doctrine\DBAL\Driver\PDOException::getSQLState()` no longer falls back to the driver error code.
+
+The method was used internally and is no longer needed.
```
Comment (Majkl578) [inline UPGRADE.md]: This shouldn't be here, looks like a copy&paste error.

File: lib/Doctrine/DBAL/Driver/DriverException.php (hunk 1 above)
Comment (Ocramius) [reply lib/Doctrine/DBAL/Driver/DriverException.php]: My mistake indeed: my snippet was missing something: https://3v4l.org/4933a

File: UPGRADE.md (hunk 2 above)
Comment (morozov) [reply UPGRADE.md]: Yes, incorrect conflict resolution.

File: lib/Doctrine/DBAL/Driver/DriverException.php (hunk 1 above)
Comment (morozov) [reply lib/Doctrine/DBAL/Driver/DriverException.php]: While it's a technically a BC break, it's not something worth noting in upgrade notes. Eventually, we'll have return type hints added across the entire codebase, so if documented explicitly, this list would become unreadably huge.

Comment (greg0ire) [reply lib/Doctrine/DBAL/Driver/DriverException.php]: That's right, a global comment on release will do

**Pull request merged**